src/main.rs

```diff
@@ -0,0 +1,289 @@
+    }
+}
+
+/// Process PPOPRF evaluation requests
+async fn randomness(
+    State(state): State<OPRFState>,
+    Json(request): Json<RandomnessRequest>,
+) -> Result<Json<RandomnessResponse>, Error> {
+    debug!("recv: {request:?}");
+    let state = state.read().map_err(|_| Error::LockFailure)?;
+    let epoch = request.epoch.unwrap_or(state.epoch);
+    if epoch != state.epoch {
+        return Err(Error::BadEpoch(epoch));
+    }
+    if request.points.len() > MAX_POINTS {
+        return Err(Error::TooManyPoints);
+    }
+    let prove = false;
+    let mut points = Vec::with_capacity(request.points.len());
+    for base64_point in request.points {
+        let input = BASE64.decode(base64_point)
+            .map_err(Error::Base64)?;
+        // FIXME: Point::from is fallible and needs to return a result.
+        // partial work-around: check correct length
+        if input.len() != ppoprf::COMPRESSED_POINT_LEN {
+            return Err(Error::BadPoint);
+        }
+        let point = ppoprf::Point::from(input.as_slice());
+        let evaluation = state.server.eval(&point, epoch, prove)
+            .map_err(Error::Oprf)?;
+        points.push(BASE64.encode(evaluation.output.as_bytes()));
+    }
+    let response = RandomnessResponse { points, epoch };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Process PPOPRF epoch and key requests
+async fn info(
+    State(state): State<OPRFState>
+) -> Result<Json<InfoResponse>, Error> {
+    debug!("recv: info reqeust");
+    let state = state.read().map_err(|_| Error::LockFailure)?;
+    let current_epoch = state.epoch;
+    // FIXME: return the end of the current epoch
+    let next_epoch_time = "unknown".to_owned();
+    let max_points = MAX_POINTS;
+    let public_key = state.server.get_public_key()
+        .serialize_to_bincode()
+        .map_err(Error::Oprf)?;
+    let public_key = BASE64.encode(public_key);
+    let response = InfoResponse {
+        current_epoch,
+        next_epoch_time,
+        max_points,
+        public_key,
+    };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Advance to the next epoch on a timer
+async fn epoch_update_loop(state: OPRFState, epochs: EpochRange) {
+    let interval = std::time::Duration::from_secs(5);
+    info!("rotating epoch every {} seconds", interval.as_secs());
+
+    loop {
+        // Wait until the current epoch ends.
+        tokio::time::sleep(interval).await;
+
+        // Acquire exclusive access to the oprf state.
+        // Panics if this fails, since processing requests with an
+        // expired epoch weakens user privacy.
+        let mut s = state.write().expect("Failed to lock OPRFState");
+
+        // Puncture the current epoch so it can no longer be used.
+        let old_epoch = s.epoch;
+        s.server.puncture(old_epoch).expect("Failed to puncture current epoch");
+
+        // Advance to the next epoch.
+        let new_epoch = old_epoch + 1;
```

Comment:
this will panic if `old_epoch` is 255, right? `old_epoch.wrapping_add(1)` should fix that

Reply:
Ah, i guess that wouldn't happen if the range ends at 255 exclusively.

Reply:
upon further review, it looks like an exclusive range would be much easier

Reply:
Addressed by 64367e1e1396a9f8b04e48c3d6ddef58380a982a.

I think the exclusive range is still harder because we have to convert to epoch values back-and-forth to u8 is more places. I went with a single overflow check instead.

Reply:
where would we have to convert?

Reply:
server.puncture and OPRFServer::new

Reply:
sorry, i'm little lost. what type would we have to convert to/from? wouldn't we still use u8 everywhere?

Reply:
I'm feeling lost too.

My thinking was: if we use an exclusive `Range<u8>` we can work with `u8` values everywhere but can't represent the value `255`, so we have to wrap after `254` which is weird. If we use a `Range<u16>` we can represent 255, but we need to cast the epoch value `as u16` when we pull it out of the server state so we can increment without overflow and test for range membership. Then we have to cast it back to u8 when we store the updated value.

Maybe you could show me what you're thinking of? Or be specific about what you object to with the current code?

Reply:
If 254 was the max value within the range, we shouldn't need to convert to u16, right? Once we increment the value to 255, `Range::contains` will return false which will recreate the OPRFServer starting at the first epoch, which means should never overflow.

Having said that, it kind of sucks that we won't be able to use the value 255 in the epoch range with my proposed solution, so I think your solution is totally fine.